pkg/logs/tail.go

```diff
@@ -0,0 +1,610 @@
+	// re-request on failure.
+	var lastBatchID string
+
+	for {
+		// Check to see if we already passed the "to" requirement.
+		if toWindow != 0 && curWindow > toWindow {
+			text.Info(out, "Reached window: %v which is newer than the requested 'to': %v", curWindow, toWindow)
+			// We are done, but we still want printing to finish.
+			close(c.doneCh)
+			break
+		}
+
+		req, err := http.NewRequest("GET", path, nil)
+		req.Header.Add("Fastly-Key", c.token)
+		if err != nil {
+			text.Error(out, "unable to create new request: %v", err)
+			os.Exit(1)
+		}
+
+		resp, err := c.doReq(req)
+		if err != nil {
+			text.Error(out, "unable to request: %v", err)
+			os.Exit(1)
+		}
+
+		// Check that our request was successful. If the server is
+		// having trouble, retry after waiting for some time.
+		if resp.StatusCode != http.StatusOK {
+			// In an effort to clean up the output, do not print on
+			// 503's.
+			if resp.StatusCode != http.StatusServiceUnavailable {
+				text.Warning(out, "non-200 resp %d", resp.StatusCode)
+			}
+			io.Copy(ioutil.Discard, resp.Body)
+			// Try the response again after a 1 second wait.
+			if resp.StatusCode/100 == 5 && resp.StatusCode != 501 ||
+				resp.StatusCode == 429 {
+				time.Sleep(1 * time.Second)
+				continue
+			}
+			// Failing at this point is unrecoverable,
+			return
+		}
+
+		// Read and parse response, send batches to the output loop.
+		scanner := bufio.NewScanner(resp.Body)
+		// Use a 10MB buffer for the bufio scanner, as we don't know
+		// how big some of the responses will be.
+		const tmb = 10 << 20
+		buf := make([]byte, tmb)
+		scanner.Buffer(buf, tmb)
+
+		for scanner.Scan() {
+			// Scan one line at a time, and get only one batch
+			// at a time.
+			b := scanner.Bytes()
+			batch, err := parseResponseData(b)
+			if err != nil {
+				// We can't parse the response, attempt to
+				// re-request from the last window & batch.
+				text.Warning(out, "unable to parse resp body: %v", err)
+				path, err = makeNewPath(path, curWindow, lastBatchID)
+				if err != nil {
+					// TODO: consider makeNewPath never returning an error?
+					text.Error(out, "error generating request URL: %v", err)
+					os.Exit(1)
+				}
+				continue
+			}
+
+			// If we got a batch back, there will be an ID.
+			if batch.ID != "" {
+				// Record last batchID in case
+				// anything fails along the way, we
+				// can re-request.
+				lastBatchID = batch.ID
+				// Send batch down batchCh to the output loop.
+				c.batchCh <- batch
+			}
+
+		}
+		resp.Body.Close()
+
+		if err := scanner.Err(); err != nil {
+			// ErrUnexpectedEOFs need to be retried, but they
+			// produce a lot of noise for the user, so don't log.
+			if err != io.ErrUnexpectedEOF {
+				text.Warning(out, "error scanning response body: %v", err)
+			}
+			// Something happened in the scanner, re-request the
+			// current batchID.
+			path, err = makeNewPath(path, curWindow, lastBatchID)
```

Comment:
This call and the subsequent error failure handling logic appears in a few places. Might be overkill but could we shift to a separate function?

Reply:
At some point I had a bunch of TODO's saying that it was annoying we did the error handling over and over.  I think I might just make `makeNewPath` never return an error.